src/client/automation/playback/visible-element-automation.js

```diff
@@ -35,66 +36,67 @@ export default class VisibleElementAutomation extends serviceUtils.EventEmitter
     }
 
     _getElementForEvent (eventArgs) {
-        var { x, y }        = eventArgs.point;
-        var expectedElement = positionUtils.containsOffset(this.element, this.options.offsetX, this.options.offsetY) ? this.element : null;
+        const { x, y }        = eventArgs.point;
+        const expectedElement = positionUtils.containsOffset(this.element, this.options.offsetX, this.options.offsetY) ? this.element : null;
 
         return getElementFromPoint(x, y, expectedElement).then(({ element }) => element);
     }
 
     _moveToElement () {
-        var moveOptions    = new MoveOptions(extend({ skipScrolling: true }, this.options), false);
-        var moveAutomation = new MoveAutomation(this.element, moveOptions);
+        const moveOptions    = new MoveOptions(extend({ skipScrolling: true }, this.options), false);
+        const moveAutomation = new MoveAutomation(this.element, moveOptions);
 
         return moveAutomation
             .run()
             .then(() => delay(this.automationSettings.mouseActionStepDelay));
     }
 
     _scrollToElement () {
-        var scrollAutomation = new ScrollAutomation(this.element, new ScrollOptions(this.options));
+        const scrollAutomation = new ScrollAutomation(this.element, new ScrollOptions(this.options));
 
         return scrollAutomation
             .run()
             .then(() => delay(this.automationSettings.mouseActionStepDelay));
     }
 
     _wrapAction (action) {
-        var offsetX                    = this.options.offsetX;
-        var offsetY                    = this.options.offsetY;
-        var screenPointBeforeAction    = getAutomationPoint(this.element, offsetX, offsetY);
-        var clientPositionBeforeAction = positionUtils.getClientPosition(this.element);
+        const roundFn                    = browserUtils.isFirefox ? Math.ceil : Math.round;
+        const offsetX                    = this.options.offsetX;
+        const offsetY                    = this.options.offsetY;
+        const screenPointBeforeAction    = getAutomationPoint(this.element, offsetX, offsetY, roundFn);
```

Comment:
Why we need to pass `roundFn` to the `getAutomationPoint`?

Reply:
`getAutomationPoint` based on hammerhead's `getOffsetPosition`. Currently the `getOffsetPosition` is implemented to return rounded X, Y offset of element. When it rounds `50.1`px to '50'px it makes `document.elementFromPoint` work wrong in FF, but if works perfectly fine in Chrome/IE. So we need to round `50.1` not to `50` but to `51` only in FF.
The test checks that that even with different positions we click in correct position of element (0, 0) in all browsers.

Reply:
Should we use these different round functions for `getAutomationPoint` only in this place? What about `drag` or `move` automation and etc.?

Reply:
Need to research, currently I'm not working on this PR

Reply:
Looks like we do not need to use different round functions in other places. The issue occurs because of combination `getAutomationPoint` and `elementFromPoint`. There is no such a combination in other places.

Reply:
Yep, let's create separate issue for IE

Reply:
https://github.com/DevExpress/testcafe/issues/2777

Reply:
I've reasearched the issue in IE, and I've found that in case of drag (or hover) we can't guarantee that target element in { x, y } will be an expected element. For example in the drag automation, for calculating the target { x, y } position we get source element coordinates with the `getAutomationPoint` method and then add { offsetX, offsetY } integer values. Depending on source element position it can be rounded down or up, i.e. 3.1 => 3 but 3.5 => 4, so target coors can differ. Besides, target element can also have fractional position, so we can not predict how we need to round the coordinates.
I've decided to fix it only in FF, because this scenario is common and very easy to reproduce.